### PR TITLE
Better sorting for completions

### PR DESCRIPTION
**Pull request by hez2010**
Addresses https://github.com/AvaloniaUI/AvaloniaVS/pull/297#issuecomment-1436324233

### CODE REVIEW

**Comment by hez2010** [comment]
@maxkatz6 It's ready for review. Please take a look.
/cc: @amwx

File: CompletionEngine/Avalonia.Ide.CompletionEngine/Completion/CompletionEngine.cs [hunk 1]
```diff
@@ -390,11 +390,26 @@ IEnumerable<string> filterNamespaces(Func<string, bool> predicate)
...
+            return new CompletionSet() { Completions = SortCompletions(completions), StartPosition = curStart };
 
         return null;
     }
 
+    private List<Completion> SortCompletions(List<Completion> completions)
+    {
+        // Group the completions based on Kind, and sort the completions for each group
+        var completionGroups = completions
+            .GroupBy(i => i.Kind, (_, compl) =>
```
Comment (hez2010) [inline CompletionEngine/Avalonia.Ide.CompletionEngine/Completion/CompletionEngine.cs:402]: We may also want to determine the priority of completion Kind.

**Comment by amwx** [comment]
Properties are now first again, but not sorted alphabetically: 
image

image


Selector is still appearing last, this one is more complicated. Typing `<Style `, the space triggers a new completion session to start, which this session lists all the controls first, and Selector is after. Continuing to type `S`, doesn't trigger a new completion session, it just sorts the existing one, which keeps the existing ordering. If you cancel that completion session (escape key) and restart it, it appears as it should.

image


I'm thinking sorting by text length isn't the way to do this (WPF/UWP doesn't do this). There's a couple other minor things that I think aren't so much sorting related as more related to providing more intelligent context-based options (related to #279 which I'm actively working on). Included in this is not suggesting classes like "DataTemplateExtension" and "ControlTemplateResult" that are part of the root problem here. Sorting just alphabetically makes sense, and I would think that should automatically place the list in the order of something like `DataTemplate` -> `DataTemplates` -> `DataTemplateExtension`

**Comment by hez2010** [comment]
I'm considering that we can build a model to predict the most used completion using some machine learning methods to help us prioritize completions in the future.

**Comment by Mrxx99** [comment]
> I'm considering that we can build a model to predict the most used completion using some machine learning methods to help us prioritize completions in the future. 

So you want ro rebuild Intellicode (for XAML)?
Maybe a simpler start would be to add attributes in Avalonia that could be set on properties/events to mark them as common for that control, e.g. Button would mark Content, Command, Click and would also inherit e.g. marked Height/Width and Margin from Control, but could also unmark inherited properties with another attribute or depriorize uncommon properties.
That way not only builtin but also custom controls and controls from libraries could be annotated to improve XAML completion for them.

**Comment by hez2010** [comment]
I've added a simple logic based on my own experience to determine the priority of completions.

File: tests/CompletionEngineTests/BasicTests.cs [hunk 2]
```diff
@@ -135,13 +135,25 @@ public void Extension_DataType_Should_Be_Completed()
...
+        public void Completions_Should_Be_Sorted()
         {
             var compl = GetCompletionsFor("<DataTemplate");
 
-            Assert.Equal(2, compl.Completions.Count);
-            Assert.Equal("DataTemplate", compl.Completions[0].InsertText);
-            Assert.Equal("DataTemplates", compl.Completions[1].InsertText);
+            Assert.Equal(3, compl.Completions.Count);
+            Assert.Equal("DataTemplate", compl.Completions[0].DisplayText);
+            Assert.Equal("DataTemplateExtensions", compl.Completions[1].DisplayText);
```
Comment (maxkatz6) [inline tests/CompletionEngineTests/BasicTests.cs:144]: I don't think DataTemplateExtensions should be visible in the completion after https://github.com/AvaloniaUI/AvaloniaVS/pull/302

Comment (hez2010) [reply tests/CompletionEngineTests/BasicTests.cs]: The branch is a bit out-of-date. I've updated the base branch to resolve this.

**Comment by hez2010** [comment]
@amwx Maybe you want to test this PR again? I think this time it should works well.

**Comment by amwx** [comment]
> @amwx Maybe you want to test this PR again? I think this time it should works well.

I did a quick test yesterday I think, and it seemed to work a lot better.

**Review by maxkatz6 — APPROVED** [review]

**Pull request merged**